# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/workers/loader/CacheLoadHandler.cpp
# Commit: 00e555156df1
# Full Hash: 00e555156df1b498565fae94f2b04357baabca24
# Author: Yulia Startsev <ystartsev@mozilla.com>
# Date: 2022-11-05 09:23:50
# Description:
#   Bug 1798667 - clear cache when finished with service worker scripts; r=asuth
#   
#   I was over eager in introducing the strong pointers to WorkerLoadContext. It turns out that
#   previously when we were cycle collecting our ScriptLoadRequests, we were also cleaning up everything
#   related to WorkerLoadContext. Also, in an attempt to fix the cancellation for workers, We
# ==============================================================================

diff -r dce99b798fd5 -r 00e555156df1 dom/workers/loader/CacheLoadHandler.cpp
--- a/dom/workers/loader/CacheLoadHandler.cpp	Fri Nov 04 14:55:16 2022 +0000
+++ b/dom/workers/loader/CacheLoadHandler.cpp	Fri Nov 04 15:04:27 2022 +0000
@@ -185,7 +185,6 @@
                                     JS::Handle<JS::Value> aValue,
                                     ErrorResult& aRv) {
   AssertIsOnMainThread();
-
   if (!aValue.isObject()) {
     FailLoaders(NS_ERROR_FAILURE);
     return;
@@ -270,8 +269,10 @@
   if (mLoadContext->mCachePromise) {
     mLoadContext->mCachePromise->MaybeReject(aRv);
   }
+
   mLoadContext->mCachePromise = nullptr;
 
+  mLoadContext->ClearCacheCreator();
   if (mLoader->IsCancelled() || !mLoadContext->mRequest) {
     return;
   }
@@ -282,6 +283,12 @@
 void CacheLoadHandler::Load(Cache* aCache) {
   AssertIsOnMainThread();
   MOZ_ASSERT(aCache);
+
+  if (mLoader->IsCancelled()) {
+    Fail(NS_BINDING_ABORTED);
+    return;
+  }
+
   MOZ_ASSERT(mLoadContext->mRequest);
 
   nsCOMPtr<nsIURI> uri;
@@ -327,10 +334,6 @@
                                         ErrorResult& aRv) {
   AssertIsOnMainThread();
 
-  if (mLoader->IsCancelled() || !mLoadContext->mRequest) {
-    return;
-  }
-
   MOZ_ASSERT(mLoadContext->mCacheStatus == WorkerLoadContext::Uncached);
   Fail(NS_ERROR_FAILURE);
 }
@@ -341,7 +344,11 @@
   AssertIsOnMainThread();
   // If we have already called 'Fail', we should not proceed. If we cancelled,
   // we should similarily not proceed.
-  if (mFailed || mLoader->IsCancelled() || !mLoadContext->mRequest) {
+  if (mFailed) {
+    return;
+  }
+  if (mLoader->IsCancelled()) {
+    Fail(NS_BINDING_ABORTED);
     return;
   }
 
@@ -425,12 +432,15 @@
       if (cacheCreator) {
         cacheCreator->DeleteCache(mLoader->GetCancelResult());
       }
+      mLoadContext->ClearCacheCreator();
       return;
     }
 
     nsresult rv = DataReceivedFromCache(
         (uint8_t*)"", 0, mChannelInfo, std::move(mPrincipalInfo),
         mCSPHeaderValue, mCSPReportOnlyHeaderValue, mReferrerPolicyHeaderValue);
+
+    mLoadContext->ClearCacheCreator();
     mLoader->OnStreamComplete(mLoadContext->mRequest, rv);
     return;
   }
@@ -490,19 +500,16 @@
     return NS_OK;
   }
 
+  if (mLoader->IsCancelled() || !mLoadContext->mRequest) {
+    Fail(NS_BINDING_ABORTED);
+    return NS_OK;
+  }
+
   MOZ_ASSERT(mLoadContext->mCacheStatus == WorkerLoadContext::ReadingFromCache);
   mLoadContext->mCacheStatus = WorkerLoadContext::Cached;
 
   MOZ_ASSERT(mPrincipalInfo);
 
-  if (mLoader->IsCancelled() || !mLoadContext->mRequest) {
-    auto cacheCreator = mLoadContext->GetCacheCreator();
-    if (cacheCreator) {
-      cacheCreator->DeleteCache(mLoader->GetCancelResult());
-    }
-    return NS_OK;
-  }
-
   nsresult rv = DataReceivedFromCache(
       aString, aStringLen, mChannelInfo, std::move(mPrincipalInfo),
       mCSPHeaderValue, mCSPReportOnlyHeaderValue, mReferrerPolicyHeaderValue);